twig engine skeleton - part II

colin-kiegel opened the pull request:
Here is some skeleton code - as a suggestion.

I stripped out my parser, lexer, loader and runtime implementations and I refactured what's left according to our previous discussion #10.
- The facade is currently called `Engine`
- The `Environment` is currently called `Setup`
- The unicorn module is currently called `engine`

So what's left is the initialization logic of Engine - i.e. this works:

```
use twig::{Engine, Setup};
let twig = Engine::new(Setup::default()).unwrap();
// ..
```

And this works too (although extension::Debug is only a dummy):

```
use twig::{Setup, Engine};
use twig::extension::Debug;

let mut setup = Setup::default()
     .set_strict_variables(true)
     .add_extension(Debug::new()).unwrap();
let engine = Engine::new(setup).unwrap();
```

And it contains some error-code definitions.

colin-kiegel committed:
twig engine skeleton

colin-kiegel commented:
Please tell me, what you think. ;-)

colin-kiegel commented:
PS: As a suggestion, I moved the tokens inside of the parser like the lexer, i.e. `twig::engine::parser::tokens` because I think they are a contract between parser and lexer - outside of the parser module they are only relevant to extensions defining token parsers. It was a little test - we can relocate to `twig::engine::tokens`. Or alternatively try to re-export some extension-stuff in `extension::api`.

colin-kiegel commented:
PPS: Since this Pull Request is still rather large - I recommend a local checkout to get a feeling for it. ;-)

colin-kiegel commented:
Shall I split it into some smaller chunks?

Nercury commented:
OK, let's continue. Some interesting things in life go in a way a bit :)

I would like to finish up the lexer before starting with parser. However, I already was studying your parser before and I think we should not try to do it like PHP does in this particular case.

Instead of having a `GenericNode` with attributes, I suggest to construct static node hierarchy. If some node hierarchy item needs to be extensible, it is still very doable, but this extension point can simply be explicit trait object.

colin-kiegel committed:
solve merge conflicts

colin-kiegel committed:
remove GenericNode

colin-kiegel commented:
np - lucky for you. ;)

Sure, doing it step by step sounds like a good plan.
- I removed attribute-related methods from `Node` trait. It now reads

```
pub trait Node : Debug {
    fn tag(&self) -> &str;
    fn position(&self) -> &Position;
    fn children(&self) -> &Vec<Box<Node>>;
    fn children_mut(&mut self) -> &mut Vec<Box<Node>>;
}
```
- I removed `GenericNode`, since it would be too early for those kind of details.

I think I will focus on the two remaining 'holes' in my lexer next: _interpolation_ and _string_ state are both `unimplemented()`. I'll try to fix that the next days and also have a look if I can reuse some of your code for that. Anyway in this pull request both lexer and parser are basically not-existent.

colin-kiegel commented:
PS: I also had a bad feeling how twigphp used the attribute array for complicated magic. I share your feeling that we should try to make this logic more explicit.

Nercury commented:
> I think I will focus on the two remaining 'holes' in my lexer next: interpolation and string state are both unimplemented()

Hey, that's implemented in my code :). Let me focus on merging my lexer into yours, and it will become magically implemented :). And I will learn your code better.

About magic and attributes - well, I found it is very much possible to get rid of it.

Take a look at `Expr` node here:
[nodes/node/expr.rs](https://github.com/Nercury/twig-rs/blob/master/src/nodes/node/expr.rs)

And here the parsing for `Expr`, matching PHP expression parser implementation:
[nodes/parser/expr.rs](https://github.com/Nercury/twig-rs/blob/master/src/nodes/parser/expr.rs)

It kind of proves that it is very much possible :)

The problem I see in Twig code, is that it tries to be very abstract, but still relies on `instanceof` calls for type checking. The problem with that, however, is that once it has these hardcoded `instanceof` checks in core, the abstraction is leaked, and there is no point in it anymore.

You might ask, what if extension wants to add a new kind of node in this hierarchy? I suggest that in this case we add another node type `Expr::Custom`. So far I had no need for it, although I admit that my parser is far from complete.

Nercury commented on src/engine/mod.rs:
```diff
@@ -0,0 +1,189 @@
+    }
+
+    /// Sets the lexer instance.
+    pub fn set_lexer(&mut self, lexer: Lexer) -> &mut Engine {
```
I know the [TwigEnvironment](https://github.com/twigphp/Twig/blob/1.x/lib/Twig/Environment.php) has these setters, but I think they don't make sense in Rust. We should always create them, by default, because they are not polymorphic, and they can't change during the `compile_template`.

colin-kiegel replied:
Agreed.

Nercury commented on src/engine/mod.rs:
```diff
@@ -0,0 +1,189 @@
+#[derive(Default, Debug)]
+pub struct Engine {
+    options: Options,
+    ext: Option<Rc<ExtensionRegistry>>,
```
Why `Rc`? Why not simply a list of extensions?

colin-kiegel replied:
In my codebase:
(a) `Engine` **owns** `Parser`, which **owns** `ExpressionParser`
(b) Both `Engine`, `Parser` and `ExpressionParser` have share direct references to the extensions (alias `ExtensionRegistry`)
=> Thus I needed shared ownership (RC) here. This adds O(1) overhead for construction and drop in my codebase.

Nercury replied:
I think it is possible to get rid of it by having Extension -> Registry -> Options pipeline.

colin-kiegel replied:
Does "->" mean "owned by"?

Nercury commented on src/engine/options.rs:
```diff
@@ -0,0 +1,84 @@
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+pub enum Charset {
+    UTF8
```
Hm, this is one of those cases where I would say "screw them!" and not support anything else ;)

colin-kiegel replied:
OK. :-)

Nercury commented on src/engine/parser/error.rs:
```diff
@@ -0,0 +1,190 @@
+#[allow(dead_code)]
+#[derive(Debug)]
+pub enum ParserErrorCode {
+    Unreachable {
```
Let's leave it to `unreachable!`: either we know what we are doing, or not ;)

colin-kiegel replied:
Yeah, I know. TwigPHP has this "Logic"-Errors - this was my rust-translation. If we ever have some code branch which "_should_ be unreachable" but without being 100% certain - this could help debugging. Like possibly race conditions? I don't know. But I agree, we don't neccessarily need it from the start, so I will remove it from the skeleton.

Nercury commented on src/engine/parser/error.rs:
```diff
@@ -0,0 +1,190 @@
+        reason: String,
+        job: job::JobDump,
+    },
+    MissingExtensions,
```
Parser can really say this? I think the most it can do is `Don'tKnowWhatToDoWithThisToken`.

colin-kiegel replied:
In my codebase:

```
impl Parser {
    pub fn new(twig: &Engine) -> Result<Parser, ParserError> {
        let ext = match twig.extensions() {
            Err(e) => return Err(ParserErrorCode::MissingExtensions
                .at(loc!())
                .caused_by(e)),
            Ok(ext) => ext
        };
 \\...
```

I guess you suggest to pass extensions to the Parser-Constructor directly? Then Engine would throw this error instead of Parser - the error code would move to Engine.

But again, a detail not relevant for the skeleton - I will remove it from the skeleton.

Nercury replied:
I am thinking the parser should define some `Options` object where it lists everything it needs, and the engine should take care of collecting/validating data from extensions and constructing this `Options` object.

Same kind of deal for Lexer.

colin-kiegel replied:
Agreed - so no errors in those `new()` functions. I like that.

PS: Hmm ... should these `Options` structures have public members, or not?
- PRO: Public members would allow composition / decomposition outside of the respective module without getter/setter-bloat.
- CON: I think public members should not be mixed with logic - i.e. we shouldn't have any `impl Options {...}` in this case.

I am asking, because I am trying to move `Setup` outside of the `Engine` module. I just noticed that I have [a mix of pub-methods and logic](https://github.com/rust-web/twig/pull/12/files#diff-6c256c09a91fb70477da0b7ab2667275R27), which I am not 100% happy with. What are your thoughts on this?

Nercury commented on src/engine/parser/error.rs:
```diff
@@ -0,0 +1,190 @@
+        job: job::JobDump,
+    },
+    MissingExtensions,
+    InvalidState {
```
This can be removed, because it has no concrete meaning.

colin-kiegel replied:
I know this is a **specific** case of "should not happen" alias `unreachable`.

But using error codes instead of panics in those weird places could really help debugging. The Lexer and Parser are just too complex too argue with 100% certainty that all "shouldn't be reachable" spots really are unreachable.

In my mind the list of error codes does not be a 1:1 list of "things that can go wrong if twig is used incorrectly" - I think it can also contain errors like "Oops, something went wrong".

Rust gives us two ways to Display those errors
- Display::fmt for production use
- Debug::fmt for debugging

I see nothing wrong to include some "He's dead Jimmy" errors with possible Debug-Info for us.

But again, I will remove it at least for now, since it is a detail.

Nercury replied:
Ah, I see what you were doing, this is generalized error in chain.

Nercury replied:
No, I suggest to leave `unreachable!` for all the obvious "should not happen" cases: for example, match regexp capture group of 2 elements, the code that handles 3 has to be unreachable, independent on user template, options or any other kind of input.

For all other cases, well, we know the error, so let's just name it.

Nercury commented on src/engine/parser/error.rs:
```diff
@@ -0,0 +1,190 @@
+        item: token::stream::Item,
+        job: job::JobDump,
+    },
+    TokenError,
```
More general `TokenError` in the same list as more specific `TokenParserError`. I think both can be specific?

colin-kiegel replied:
I'll remove it for now.

Nercury replied:
Nevermind, got it. Don't agree, but got it.

Nercury commented on src/engine/parser/error.rs:
```diff
@@ -0,0 +1,190 @@
+        error: String,
+        job: job::JobDump,
+    },
+    SemanticError,
```
I am scratching my head :)

colin-kiegel replied:
I'll remove it for now.

Nercury replied:
Scratching ended.

Nercury commented on src/engine/parser/lexer/error.rs:
```diff
@@ -0,0 +1,188 @@
+    Unreachable {
+        reason: String
+    },
+    MissingExtensions,
```
I think there is no need for lexer to know that.

colin-kiegel replied:
See Parser. I'll remove it for now.

Nercury commented on src/engine/parser/lexer/error.rs:
```diff
@@ -0,0 +1,188 @@
+
+#[derive(Debug, PartialEq)]
+pub enum LexerErrorCode {
+    Unreachable {
```
Same - if it's unreachable, well, the code should not go there. Let's use `unreachable!`.

colin-kiegel replied:
I'll remove it for now.

Nercury commented on src/engine/parser/lexer/error.rs:
```diff
@@ -0,0 +1,188 @@
+        value: String
+    },
+    _InvalidState,
+    SyntaxError,
```
Almost every error is syntax error. I will look into this.

Nercury replied:
Rusty way would be `Syntax(SyntaxError)`. But then it's not suitable for `LexerErrorCode` anymore.

colin-kiegel replied:
I'll remove it for now. I guess not many error codes will be left after this cleanup... ;-)

Nercury commented on src/engine/parser/lexer/job/cursor.rs:
```diff
@@ -0,0 +1,36 @@
+
+pub type Position = usize;
+pub type Line = usize;
+pub type CursorDump = String;
```
Unstructured data :/

colin-kiegel replied:
True, I just took a lazy shortcut here. I will have another look.

Nercury replied:
I suggested just build structured errors without abstraction.

Nercury commented:
By the way, this is going to clash horribly with the lexer skeleton :/

Also, I want to leave errors ungeneralized, especially bubble up the `SyntaxError` to top as _the most important error you can receive from twig_.

I would like the hierarchy _in single crate boundaries_ to be done [in very boring and obvious way](https://doc.rust-lang.org/book/error-handling.html#composing-custom-error-types). Enums.

Of course, we will inevitably have top master `TwigError` that we will have to expose to external users, that's where we will implement the `std::error::Error` trait.

Also the trace with locations in rust source code is cool, but I very much doubt the usefulness of that. Of course, that's my opinion.

Here we go, said all the things that were piling up :)

colin-kiegel commented:
I understand the bubbling-up-thing. I will try to rework the error-Display-representation to (at least) mimic this behaviour. Still I find the location trace more than useful. I am not the kind of guy that likes stepping forth and back with a debugger and setting breakpoints. Instead I want fast and rich debug information. I know it's not for free - its like an investment and will slow down initial implementation a bit - but the return will be way faster debugging. I think it will enable us to _understand_ and fix some errors solely based on Debug::fmt on errors instead of cumbersome local debugging. I would still like to hold on to that, if we can achieve a nice representation for _users_ too.

Nercury commented:
I get that. I was doing the same using logs.

How about we implement tracing as a layer on any error, so there is no need for trait objects. For example, turn `Error<T>` into struct like this:

``` rust
/// Runtime error with stack trace.
#[derive(Clone, Debug)]
pub struct Traced<T: Clone, Debug> {
    pub error: T,
    /// Trace entries which may have location and/or another error. Maybe another struct.
    pub trace: Vec<(Location, Option<T>)>,
}

/// Implement display where T: Display
/// Maybe implement From<Traced<OtherT>> where T: From<OtherT>
```

The difference here would be that all entries in `trace` would probably have to be converted into `T`, which would be publicly visible "main" TwigError.

colin-kiegel committed:
refactor - move engine::Setup to setup::Setup

colin-kiegel committed:
refactor: move to api::{lexer, node, parser, token}

colin-kiegel committed:
refactor Extension + ExtensionRegistry

colin-kiegel committed:
remove engine::options::Charset

colin-kiegel committed:
minor cleanup

colin-kiegel commented:
Ok, here are some changes:
- refactor Extension + ExtensionRegistry (**tell don't ask**)
- move to api::{lexer, node, parser, token}
- move engine::Setup to setup::Setup
- some cleanup (like removing some error codes, option::Charset, etc.)

> You might ask, what if extension wants to add a new kind of node in this hierarchy? I suggest that in this case we add another node type Expr::Custom. So far I had no need for it, although I admit that my parser is far from complete.

In my mind extensions define _most_ node types by themselves, so we need to define what we _expect_ those nodes to implement. Even `if` and `for``nodes could just be defined by the`core` extension. We could at least try, if that works.

Nercury commented:
> so we need to define what we expect those nodes to implement.

I expect them to implement:
- Abstract extension, which configures what is available for lexer and parser;
- Abstract token parser, which transforms token stream into node tree.
- Function evaluations, so they can be dispatched to extensions at runtime (for functions, filters, operators).

The node tree will be evaluated, somehow. If the base collection of nodes is enough for extension to achieve what's needed, we don't need abstract nodes here.

The issue arises because some extension might want to add something to node tree that can not be evaluated by base collection of nodes. In this case the evaluation will need to be provided by extension anyways, so I would argue that in such case an extension can tell explicitly what the evaluation is in the form of CustomNode.

I imagine two routes for this:
- `CustomNode<Box<Evaluate>>`. Here, the extension provides a way to evaluate this node.
- `CustomNode { id: u32 }`. Here, the extension must provide evaluation for id the same way it provides token parsers for tokens by token name.

I think the second way is more flexible, because the extension may implement multiple ways to evaluate node for different backends.

I hope I managed to show here that abstraction can be done in different ways and putting it in the center (Node) is not necessary.

colin-kiegel committed:
merged master

Nercury commented:
Continued in #20.
